File: rust/crates/dorado-engine/src/format.rs
//! On-disk container for password-encrypted output.
//!
//! All multi-byte integers are big-endian and every field is byte-aligned (no
//! bit packing). The layout is:
//!
//! ```text
//! magic "DRDO" (4) | version (1) | variant (1) | kdf id (1) |
//! kdf params (fixed per kdf) | mac id (1) | chunk size (u32) |
//! salt len (1) | salt | tweak (16) | iv (block size) | frames...
//! ```
//!
//! The body after the header is a sequence of authenticated frames (see `mac`
//! and the CLI streaming code); each frame carries its own HMAC tag, and the
//! header is bound into the first frame's tag. Everything in the header is
//! non-secret. A tampered header, a tampered or reordered frame, a wrong
//! password, or a truncated stream is detected on decryption.

use std::io::Read;

use crate::kdf::{KdfParams, PrfId};

pub const MAGIC: &[u8; 4] = b"DRDO";
/// Current container format version. Version 3 (no label field) is still read.
pub const VERSION: u8 = 4;
/// Upper bound on the optional label, to bound allocation when reading a header.
pub const MAX_LABEL_LEN: usize = 4096;

/// Which MAC authenticates the file. Stored as a byte so the set can grow. All
/// three produce a 32-byte tag from the 32-byte MAC key.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum MacId {
    /// HMAC-SHA256.
    HmacSha256,
    /// Skein-512 in its keyed (MAC) mode. The default.
    Skein512,
    /// BLAKE3 in keyed mode.
    Blake3,
}

impl MacId {
    /// The on-disk byte identifying this MAC.
    pub fn code(self) -> u8 {
        match self {
            MacId::HmacSha256 => 1,
            MacId::Skein512 => 2,
            MacId::Blake3 => 3,
        }
    }

    /// Parse a MAC id byte, rejecting unknown values.
    pub fn from_code(b: u8) -> Result<Self, String> {
        match b {
            1 => Ok(MacId::HmacSha256),
            2 => Ok(MacId::Skein512),
            3 => Ok(MacId::Blake3),
            n => Err(format!("unknown mac id {n}")),
        }
    }
}

/// Which Threefish block size a file uses. The code byte is stored in the
/// header and also fixes the key and IV lengths.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Variant {
    /// Threefish-256 (32-byte block, key, and IV).
    T256,
    /// Threefish-512 (64-byte block, key, and IV).
    T512,
    /// Threefish-1024 (128-byte block, key, and IV).
    T1024,
}

impl Variant {
    /// The on-disk byte identifying this variant.
    pub fn code(self) -> u8 {
        match self {
            Variant::T256 => 0,
            Variant::T512 => 1,
            Variant::T1024 => 2,
        }
    }

    /// Parse a variant code byte, rejecting unknown values.
    pub fn from_code(b: u8) -> Result<Self, String> {
        match b {
            0 => Ok(Variant::T256),
            1 => Ok(Variant::T512),
            2 => Ok(Variant::T1024),
            n => Err(format!("unknown variant code {n}")),
        }
    }

    /// Key length in bytes. The block and IV are the same size.
    pub fn key_len(self) -> usize {
        match self {
            Variant::T256 => 32,
            Variant::T512 => 64,
            Variant::T1024 => 128,
        }
    }

    /// Block length in bytes (equal to the key and IV length).
    pub fn block_len(self) -> usize {
        self.key_len()
    }
}

/// The non-secret header that precedes the ciphertext in a password file.
#[derive(Debug)]
pub struct Header {
    /// On-disk format version this header was written as. New headers use
    /// [`VERSION`]; a header read from a v3 file keeps `3` so `to_bytes`
    /// reproduces the original bytes (the header is bound into chunk 0's tag).
    pub version: u8,
    pub variant: Variant,
    pub kdf: KdfParams,
    pub mac: MacId,
    /// Plaintext bytes per authenticated frame. A multiple of the block size.
    pub chunk_size: u32,
    pub salt: Vec<u8>,
    pub tweak: [u8; 16],
    pub iv: Vec<u8>,
    /// Optional non-secret label bound into the authentication (version 4+).
    /// Empty when absent or for v3 files.
    pub label: Vec<u8>,
}

impl Header {
    /// Serialize the header to its on-disk bytes (ciphertext is appended by the
    /// caller).
    pub fn to_bytes(&self) -> Vec<u8> {
        let mut out = Vec::new();
        out.extend_from_slice(MAGIC);
        out.push(self.version);
        out.push(self.variant.code());
        match &self.kdf {
            KdfParams::Argon2id {
                m_cost,
                t_cost,
                p_cost,
            } => {
                out.push(1);
                out.extend_from_slice(&m_cost.to_be_bytes());
                out.extend_from_slice(&t_cost.to_be_bytes());
                out.extend_from_slice(&p_cost.to_be_bytes());
            }
            KdfParams::Scrypt { log_n, r, p } => {
                out.push(2);
                out.push(*log_n);
                out.extend_from_slice(&r.to_be_bytes());
                out.extend_from_slice(&p.to_be_bytes());
            }
            KdfParams::Pbkdf2 { rounds, prf } => {
                out.push(3);
                out.extend_from_slice(&rounds.to_be_bytes());
                out.push(prf.code());
            }
        }
        out.push(self.mac.code());
        out.extend_from_slice(&self.chunk_size.to_be_bytes());
        // Salt length fits in a byte by construction (we generate 16).
        out.push(self.salt.len() as u8);
        out.extend_from_slice(&self.salt);
        out.extend_from_slice(&self.tweak);
        out.extend_from_slice(&self.iv);
        // Version 4+ appends the optional label at the end, so the v3 prefix is
        // byte-identical and v3 headers reserialize exactly.
        if self.version >= 4 {
            out.extend_from_slice(&(self.label.len() as u16).to_be_bytes());
            out.extend_from_slice(&self.label);
        }
        out
    }

    /// Read and parse a header from a stream, leaving the reader positioned at
    /// the first frame.
    pub fn read<R: Read + ?Sized>(r: &mut R) -> Result<Header, String> {
        if take(r, 4, "magic")? != MAGIC {
            return Err("not a dorado password file (bad magic)".into());
        }
        let version = u8r(r, "version")?;
        if version != 3 && version != VERSION {
            return Err(format!("unsupported format version {version}"));
        }
        let variant = Variant::from_code(u8r(r, "variant")?)?;
        let kdf = match u8r(r, "kdf id")? {
            1 => KdfParams::Argon2id {
                m_cost: u32r(r, "m_cost")?,
                t_cost: u32r(r, "t_cost")?,
                p_cost: u32r(r, "p_cost")?,
            },
            2 => KdfParams::Scrypt {
                log_n: u8r(r, "log_n")?,
                r: u32r(r, "r")?,
                p: u32r(r, "p")?,
            },
            3 => KdfParams::Pbkdf2 {
                rounds: u32r(r, "rounds")?,
                prf: PrfId::from_code(u8r(r, "prf")?)?,
            },
            n => return Err(format!("unknown kdf id {n}")),
        };
        let mac = MacId::from_code(u8r(r, "mac id")?)?;
        let chunk_size = u32r(r, "chunk size")?;
        let salt_len = u8r(r, "salt len")? as usize;
        let salt = take(r, salt_len, "salt")?;
        let mut tweak = [0u8; 16];
        tweak.copy_from_slice(&take(r, 16, "tweak")?);
        let iv = take(r, variant.block_len(), "iv")?;
        // The label is present only in version 4+.
        let label = if version >= 4 {
            let label_len = u16r(r, "label len")? as usize;
            if label_len > MAX_LABEL_LEN {
                return Err(format!("label too long ({label_len} bytes)"));
            }
            take(r, label_len, "label")?
        } else {
            Vec::new()
        };
        Ok(Header {
            version,
            variant,
            kdf,
            mac,
            chunk_size,
            salt,
            tweak,
            iv,
            label,
        })
    }

    /// Parse a header from the start of `data`, returning it and the offset at
    /// which the frames begin. Convenience wrapper over `read` for in-memory use.
    #[cfg(test)]
    pub fn parse(data: &[u8]) -> Result<(Header, usize), String> {
        let mut cur = std::io::Cursor::new(data);
        let header = Header::read(&mut cur)?;
        Ok((header, cur.position() as usize))
    }
}

/// Read exactly `n` bytes, mapping a short read to a descriptive error.
fn take<R: Read + ?Sized>(r: &mut R, n: usize, what: &str) -> Result<Vec<u8>, String> {
    let mut buf = vec![0u8; n];
    r.read_exact(&mut buf)
        .map_err(|_| format!("header truncated reading {what}"))?;
    Ok(buf)
}

fn u8r<R: Read + ?Sized>(r: &mut R, what: &str) -> Result<u8, String> {
    Ok(take(r, 1, what)?[0])
}

fn u16r<R: Read + ?Sized>(r: &mut R, what: &str) -> Result<u16, String> {
    let b = take(r, 2, what)?;
    Ok(u16::from_be_bytes([b[0], b[1]]))
}

fn u32r<R: Read + ?Sized>(r: &mut R, what: &str) -> Result<u32, String> {
    let b = take(r, 4, what)?;
    Ok(u32::from_be_bytes([b[0], b[1], b[2], b[3]]))
}

#[cfg(test)]
mod tests;

File: java/src/main/java/com/alleato/dorado/engine/Format.java
package com.alleato.dorado.engine;

import static java.nio.charset.StandardCharsets.US_ASCII;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;

/**
 * The on-disk container format (magic "DRDO", version 4; version 3 still read):
 * header marshalling and streaming parse. Matches the Rust/Go/TS ports
 * byte-for-byte. All multi-byte integers are big-endian. {@code u32} header fields
 * are held in {@code long} so the validation bounds (in {@link Kdf}) cannot be
 * bypassed by a hostile value with the high bit set.
 */
public final class Format {
    public static final String MAGIC = "DRDO";
    public static final int VERSION = 4;
    public static final int DEFAULT_CHUNK_BYTES = 64 * 1024;
    /** Hard ceiling on an accepted chunk size. The effective cap can only be lower. */
    public static final long MAX_CHUNK_BYTES = 1L << 30;
    /** Default accepted chunk-size cap when the environment does not override it. */
    public static final long DEFAULT_MAX_CHUNK_BYTES = 64L * 1024 * 1024;
    public static final int MAX_LABEL_LEN = 4096;
    public static final int MAC_KEY_LEN = 32;
    public static final int TAG_LEN = 32;

    // Variant codes.
    public static final int T256 = 0;
    public static final int T512 = 1;
    public static final int T1024 = 2;

    // MAC ids.
    public static final int MAC_HMAC = 1;
    public static final int MAC_SKEIN = 2;
    public static final int MAC_BLAKE3 = 3;

    // KDF ids.
    public static final int KDF_ARGON2ID = 1;
    public static final int KDF_SCRYPT = 2;
    public static final int KDF_PBKDF2 = 3;
    public static final int PRF_HMAC_SHA256 = 1;

    private Format() {}

    /**
     * The effective accepted chunk-size cap. Reads the {@code DORADO_MAX_CHUNK_BYTES}
     * environment variable and clamps it into {@code [1, MAX_CHUNK_BYTES]}; an unset,
     * empty, or unparseable value yields {@link #DEFAULT_MAX_CHUNK_BYTES}. The override
     * can only tighten the cap below the hard ceiling, never raise it past 1 GiB.
     */
    public static long maxChunkBytes() {
        return clampChunkCap(System.getenv("DORADO_MAX_CHUNK_BYTES"));
    }

    /**
     * Pure helper backing {@link #maxChunkBytes()}: maps a raw env-var string to an
     * effective chunk-size cap. {@code null}, empty, or non-parseable-as-a-non-negative
     * integer yields {@link #DEFAULT_MAX_CHUNK_BYTES}; a parsed value is clamped into
     * {@code [1, MAX_CHUNK_BYTES]}.
     */
    public static long clampChunkCap(String raw) {
        if (raw == null || raw.isEmpty()) {
            return DEFAULT_MAX_CHUNK_BYTES;
        }
        long v;
        try {
            v = Long.parseLong(raw.trim());
        } catch (NumberFormatException e) {
            return DEFAULT_MAX_CHUNK_BYTES;
        }
        if (v < 0) {
            return DEFAULT_MAX_CHUNK_BYTES;
        }
        if (v < 1) {
            return 1;
        }
        return Math.min(v, MAX_CHUNK_BYTES);
    }

    /** The variant's key/block/IV length in bytes. */
    public static int keyLen(int variant) {
        return switch (variant) {
            case T256 -> 32;
            case T512 -> 64;
            case T1024 -> 128;
            default -> throw new IllegalArgumentException("unknown variant code " + variant);
        };
    }

    public static int blockLen(int variant) {
        return keyLen(variant);
    }

    /** A KDF choice with its cost parameters (only the fields for {@code kind} are used). */
    public static final class KdfParams {
        public final int kind;
        public long mCost;  // argon2id: memory in KiB
        public long tCost;  // argon2id: iterations
        public long pCost;  // argon2id: lanes
        public int logN;    // scrypt: log2(N)
        public long r;      // scrypt: block factor
        public long p;      // scrypt: parallelism
        public long rounds; // pbkdf2: iterations
        public int prf;     // pbkdf2: PRF id

        public KdfParams(int kind) {
            this.kind = kind;
        }

        public static KdfParams argon2id(long mCostKib, long tCost, long pCost) {
            KdfParams k = new KdfParams(KDF_ARGON2ID);
            k.mCost = mCostKib;
            k.tCost = tCost;
            k.pCost = pCost;
            return k;
        }

        public static KdfParams scrypt(int logN, long r, long p) {
            KdfParams k = new KdfParams(KDF_SCRYPT);
            k.logN = logN;
            k.r = r;
            k.p = p;
            return k;
        }

        public static KdfParams pbkdf2(long rounds) {
            KdfParams k = new KdfParams(KDF_PBKDF2);
            k.rounds = rounds;
            k.prf = PRF_HMAC_SHA256;
            return k;
        }
    }

    /** The non-secret header preceding the ciphertext in a password file. */
    public static final class Header {
        public int version;
        public int variant;
        public KdfParams kdf;
        public int mac;
        public long chunkSize;
        public byte[] salt;
        public byte[] tweak; // 16 bytes
        public byte[] iv;
        public byte[] label;
    }

    private static void putU16(ByteArrayOutputStream b, int v) {
        b.write((v >>> 8) & 0xff);
        b.write(v & 0xff);
    }

    private static void putU32(ByteArrayOutputStream b, long v) {
        b.write((int) ((v >>> 24) & 0xff));
        b.write((int) ((v >>> 16) & 0xff));
        b.write((int) ((v >>> 8) & 0xff));
        b.write((int) (v & 0xff));
    }

    /** Serialize a header to its on-disk bytes. */
    public static byte[] marshal(Header h) {
        ByteArrayOutputStream b = new ByteArrayOutputStream();
        b.writeBytes(MAGIC.getBytes(US_ASCII));
        b.write(h.version);
        b.write(h.variant);
        switch (h.kdf.kind) {
            case KDF_ARGON2ID -> {
                b.write(1);
                putU32(b, h.kdf.mCost);
                putU32(b, h.kdf.tCost);
                putU32(b, h.kdf.pCost);
            }
            case KDF_SCRYPT -> {
                b.write(2);
                b.write(h.kdf.logN);
                putU32(b, h.kdf.r);
                putU32(b, h.kdf.p);
            }
            case KDF_PBKDF2 -> {
                b.write(3);
                putU32(b, h.kdf.rounds);
                b.write(h.kdf.prf);
            }
            default -> throw new IllegalArgumentException("unknown kdf kind " + h.kdf.kind);
        }
        b.write(h.mac);
        putU32(b, h.chunkSize);
        b.write(h.salt.length);
        b.writeBytes(h.salt);
        b.writeBytes(h.tweak);
        b.writeBytes(h.iv);
        if (h.version >= 4) {
            putU16(b, h.label.length);
            b.writeBytes(h.label);
        }
        return b.toByteArray();
    }

    private static byte[] readExact(InputStream in, int n, String what) throws IOException {
        byte[] b = in.readNBytes(n);
        if (b.length != n) {
            throw new DoradoException("header truncated reading " + what);
        }
        return b;
    }

    private static int readU8(InputStream in, String what) throws IOException {
        return readExact(in, 1, what)[0] & 0xff;
    }

    private static int readU16(InputStream in, String what) throws IOException {
        byte[] b = readExact(in, 2, what);
        return (b[0] & 0xff) << 8 | (b[1] & 0xff);
    }

    private static long readU32(InputStream in, String what) throws IOException {
        byte[] b = readExact(in, 4, what);
        return (b[0] & 0xffL) << 24 | (b[1] & 0xffL) << 16 | (b[2] & 0xffL) << 8 | (b[3] & 0xffL);
    }

    /** Read a header from in, leaving it positioned at the first frame. */
    public static Header read(InputStream in) throws IOException {
        byte[] m = readExact(in, 4, "magic");
        if (!new String(m, US_ASCII).equals(MAGIC)) {
            throw new DoradoException("not a dorado password file (bad magic)");
        }
        Header h = new Header();
        h.version = readU8(in, "version");
        if (h.version != 3 && h.version != VERSION) {
            throw new DoradoException("unsupported format version " + h.version);
        }
        h.variant = readU8(in, "variant");
        keyLen(h.variant); // validates the code

        int kdfId = readU8(in, "kdf id");
        KdfParams kdf;
        switch (kdfId) {
            case KDF_ARGON2ID -> {
                kdf = new KdfParams(KDF_ARGON2ID);
                kdf.mCost = readU32(in, "m_cost");
                kdf.tCost = readU32(in, "t_cost");
                kdf.pCost = readU32(in, "p_cost");
            }
            case KDF_SCRYPT -> {
                kdf = new KdfParams(KDF_SCRYPT);
                kdf.logN = readU8(in, "log_n");
                kdf.r = readU32(in, "r");
                kdf.p = readU32(in, "p");
            }
            case KDF_PBKDF2 -> {
                kdf = new KdfParams(KDF_PBKDF2);
                kdf.rounds = readU32(in, "rounds");
                int prf = readU8(in, "prf");
                if (prf != PRF_HMAC_SHA256) {
                    throw new DoradoException("unknown prf id " + prf);
                }
                kdf.prf = prf;
            }
            default -> throw new DoradoException("unknown kdf id " + kdfId);
        }
        h.kdf = kdf;

        h.mac = readU8(in, "mac id");
        if (h.mac < 1 || h.mac > 3) {
            throw new DoradoException("unknown mac id " + h.mac);
        }
        h.chunkSize = readU32(in, "chunk size");
        int saltLen = readU8(in, "salt len");
        h.salt = readExact(in, saltLen, "salt");
        h.tweak = readExact(in, 16, "tweak");
        h.iv = readExact(in, blockLen(h.variant), "iv");
        if (h.version >= 4) {
            int labelLen = readU16(in, "label len");
            if (labelLen > MAX_LABEL_LEN) {
                throw new DoradoException("label too long (" + labelLen + " bytes)");
            }
            h.label = readExact(in, labelLen, "label");
        } else {
            h.label = new byte[0];
        }
        return h;
    }
}
